lib/mailbox.c

```diff
@@ -16,25 +16,116 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_debug.h>
+#include <rte_malloc.h>
+
 #include "gatekeeper_mailbox.h"
 
-/*
- * TODO Remove this preprocessing directive when there are
- * are calls to these functions; for now, the mailbox is
- * just an example of how to set up a library.
- */
-#if 0
-void
-create_mailbox(void)
+/* XXX Sample parameters, need to be tested for better performance. */
+#define GK_MEM_CACHE_SIZE (64)
+
+struct mailbox *
+create_mailbox(const char *tag,
+	int ele_count, int ele_size, unsigned int lcore_id)
+{
+	int ret;
+	char ring_name[64];
+	char pool_name[64];
+	char mailbox_name[64];
+	struct mailbox *mb = NULL;
+	unsigned int socket_id = rte_lcore_to_socket_id(lcore_id);
+
+	ret = snprintf(mailbox_name,
+		sizeof(mailbox_name), "%s_mailbox_%u", tag, lcore_id);
+	RTE_ASSERT(ret < sizeof(mailbox_name));
+
+	mb = (struct mailbox *)rte_calloc(mailbox_name, 1, sizeof(*mb), 0);
+	if (mb == NULL) {
+		RTE_LOG(ERR, MALLOC,
+			"mailbox: can't create mailbox %s (len = %d) at lcore %u!\n",
+			mailbox_name, ret, lcore_id);
+		return NULL;
+	}
+
+	ret = snprintf(ring_name,
+		sizeof(ring_name), "%s_ring_%u", tag, lcore_id);
```

Comment:
Please change string `"%s_ring_%u"` to `"%s_mailbox_ring_%u"` since it's not any ring.